src/Libraries/Microsoft.Extensions.AI/ChatCompletion/FunctionInvokingChatClient.cs

```diff
@@ -552,75 +554,88 @@ private static void UpdateOptionsForNextIteration(ref ChatOptions options, strin
 
             if (AllowConcurrentInvocation)
             {
-                // Rather than await'ing each function before invoking the next, invoke all of them
+                // Rather than awaiting each function before invoking the next, invoke all of them
                 // and then await all of them. We avoid forcibly introducing parallelism via Task.Run,
                 // but if a function invocation completes asynchronously, its processing can overlap
                 // with the processing of other the other invocation invocations.
-                results = await Task.WhenAll(
-                    from i in Enumerable.Range(0, functionCallContents.Count)
+                results.AddRange(await Task.WhenAll(
+                    from callIndex in Enumerable.Range(0, functionCallContents.Count)
                     select ProcessFunctionCallAsync(
                         messages, options, functionCallContents,
-                        iteration, i, captureExceptions: true, cancellationToken));
+                        iteration, callIndex, captureExceptions: true, isStreaming, cancellationToken)));
+
+                shouldTerminate = results.Any(r => r.Terminate);
             }
             else
             {
                 // Invoke each function serially.
-                results = new FunctionInvocationResult[functionCallContents.Count];
-                for (int i = 0; i < results.Length; i++)
+                for (int callIndex = 0; callIndex < functionCallContents.Count; callIndex++)
                 {
-                    results[i] = await ProcessFunctionCallAsync(
+                    var functionResult = await ProcessFunctionCallAsync(
                         messages, options, functionCallContents,
-                        iteration, i, captureCurrentIterationExceptions, cancellationToken);
-                }
-            }
-
-            var shouldTerminate = false;
+                        iteration, callIndex, captureCurrentIterationExceptions, isStreaming, cancellationToken);
 
-            IList<ChatMessage> added = CreateResponseMessages(results);
-            ThrowIfNoFunctionResultsAdded(added);
-            UpdateConsecutiveErrorCountOrThrow(added, ref consecutiveErrorCount);
+                    results.Add(functionResult);
 
-            messages.AddRange(added);
-            foreach (FunctionInvocationResult fir in results)
-            {
-                shouldTerminate = shouldTerminate || fir.ShouldTerminate;
+                    // If any function requested termination, we should stop right away.
+                    if (functionResult.Terminate)
+                    {
+                        shouldTerminate = true;
+                        break;
+                    }
+                }
             }
 
-            return (shouldTerminate, consecutiveErrorCount, added);
+            IList<ChatMessage> addedMessages = CreateResponseMessages(results.ToArray());
```

Comment:
We don't need to block this PR on it. For the future, this ToArray could be replaced with CollectionsMarshal.AsSpan(results) under `#if NET`